### PR TITLE
Merge analyze and write phases in Painless "user" tree

### PR DESCRIPTION
**Pull request by jdconrad**
This is the next step in producing immutable state in the Painless "user" tree. This combines the analyze phase and the write phase so that both semantic checking and the ir tree are generated in the same phase. This allows the removal of the Input/Output objects created in the previous PRs from being mutable state on the nodes. Instead they are now local state for the analyze phase.

Relates: #53561 
Relates: #49869

### CODE REVIEW

**Comment by jdconrad** [comment]
@elasticmachine test elasticsearch-ci/default-distro

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EListInit.java [hunk 1]
```diff
@@ -29,76 +29,75 @@
...
-        return output;
-    }
-
-    @Override
-    ListInitializationNode write(ClassNode classNode) {
         ListInitializationNode listInitializationNode = new ListInitializationNode();
 
-        for (AExpression value : values) {
-            listInitializationNode.addArgumentNode(value.cast(value.write(classNode)));
+        for (int value = 0; value < values.size(); ++value) {
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EListInit.java]: Consider changing this to `i`.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ENewArray.java [hunk 2]
```diff
@@ -60,34 +61,34 @@ Output analyze(ScriptRoot scriptRoot, Scope scope, Input input) {
...
-        return output;
-    }
-
-    @Override
-    NewArrayNode write(ClassNode classNode) {
         NewArrayNode newArrayNode = new NewArrayNode();
 
-        for (AExpression argument : arguments) {
-            newArrayNode.addArgumentNode(argument.cast(argument.write(classNode)));
+        for (int argument = 0; argument < arguments.size(); ++ argument) {
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ENewArray.java]: `i`

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ENewArrayFunctionRef.java [hunk 3]
```diff
@@ -34,11 +35,11 @@
...
  */
-public final class ENewArrayFunctionRef extends AExpression implements ILambda {
-    private final String type;
+public class ENewArrayFunctionRef extends AExpression implements ILambda {
 
-    private SFunction function;
-    private FunctionRef ref;
+    protected final String type;
+
+    // TODO: make local
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ENewArrayFunctionRef.java]: If this isn't gonna get fixed right away, consider cutting an issue to address this.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ENewArrayFunctionRef.java [hunk 4]
```diff
@@ -48,17 +49,18 @@ public ENewArrayFunctionRef(Location location, String type) {
     }
 
     @Override
-    Output analyze(ScriptRoot scriptRoot, Scope scope, Input input) {
-        this.input = input;
-        output = new Output();
+    Output analyze(ClassNode classNode, ScriptRoot scriptRoot, Scope scope, Input input) {
+        FunctionRef ref;
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ENewArrayFunctionRef.java]: Consider moving this closer to assignment.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ENumeric.java [hunk 5]
```diff
@@ -23,20 +23,17 @@
...
  * Represents a non-decimal numeric constant.
  */
-public final class ENumeric extends AExpression {
+public class ENumeric extends AExpression {
 
-    private final String value;
-    private int radix;
-
-    protected Object constant;
+    protected final String value;
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ENumeric.java]: If these aren't going to be the extension point, no need to change visibility.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/PBrace.java [hunk 6]
```diff
@@ -34,11 +34,12 @@
...
  * Represents an array load/store and defers to a child subnode.
  */
-public final class PBrace extends AStoreable {
+public class PBrace extends AStoreable {
 
-    private AExpression index;
+    protected final AExpression index;
 
-    private AStoreable sub = null;
+    // TODO: make local
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/PBrace.java]: Punt to issue.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SFunction.java [hunk 7]
```diff
@@ -121,7 +118,8 @@ void generateSignature(PainlessLookup painlessLookup) {
                 PainlessLookupUtility.typeToJavaType(returnType), paramClasses).toMethodDescriptorString());
     }
 
-    void analyze(ScriptRoot scriptRoot) {
+    // TODO: decouple from ir
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SFunction.java]: Punt to issue.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EListInit.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EListInit.java]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ENewArray.java (hunk 2 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ENewArray.java]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ENumeric.java (hunk 5 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ENumeric.java]: If it's all right I would like to leave this for now as extending the nodes is currently the only way to do an extension point in the user tree. As other design ideas are explored, it seems more appropriate to modify them then.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ENewArrayFunctionRef.java (hunk 3 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ENewArrayFunctionRef.java]: Added issue.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ENewArrayFunctionRef.java (hunk 4 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ENewArrayFunctionRef.java]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/PBrace.java (hunk 6 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/PBrace.java]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SFunction.java (hunk 7 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SFunction.java]: Removed as this TODO no longer makes sense.

**Comment by jdconrad** [comment]
@stu-elastic Thanks for the review. I've responded to all PR comments, so I think this is ready for the next round.

**Review by stu-elastic — APPROVED** [review]
Looking good.

**Comment by jdconrad** [comment]
@stu-elastic Thanks again. Will merge as soon as CI passes.

**Pull request merged**